### DISCUSSION TITLE
Support Plugins for datasource / panel

### DISCUSSION THREAD

**Discussion by Nexucis** [perses/perses#72] [Design Discussions]
Hello there,

At some point, we would like to support different datasource than Prometheus, like ElasticSearch or MySQL. That doesn't necessary mean that the core will include that.

For example, we could think to have a dedicated repository for the plugins itself (more opened for external maintainer ?)

Usage of a different datasources induces supporting different query language, different kind of variables, perhaps different panels ?

I think even before talking about a technical solution, it will be good to :
* clarify the feature itself
* to scope the feature. Like cool be interesting to say what we won't support for example. (I supposed it will come with a technical solution, but probably we can already see things we don't want to support)

We can also start to define what "plugin" mean for us. Probably it's already clear for everybody, but we could be surprised.

On my side, a "good" plugin would be something we can add without modifying the binary (backend + frontend).

**Comment by Nexucis** [comment]
Also if it takes time to find how to do it, it can be ok to say this design will be for the v1.
We are creating an alpha version, so it's fine to break things and try it until we are most of us happy with it :smiley:. Specially for the beginning.

Even after creating the mvp, I'm expecting we will figure out that things are odd, specially after the first feedbacks. We shouldn't block ourself if we want to break things.

**Comment by LukeTillman** [comment]
Totally agree on the "it's OK to break things", especially when it comes to any plugin APIs we create. I think we'll just need to make it clear to anyone else that's interested that "this is very unstable, so you probably don't want to use it (yet)". As for what should be a plugin, on the front-end I was thinking of three pieces of functionality for the MVP:

1. Panels (i.e. visualizations)
2. Variables: currently in the UI, I'm thinking of this as "how do I get the available values for a variable so a user can make a selection?"
3. Querying for time series data (i.e. how do I get the data to show in a Panel)

Like we were discussing yesterday, I really like the idea of users having the option of doing some types of plugins on either the server or in the browser. Obviously 1 is going to be limited to client-side, but I think 2 and 3 are definitely candidates for server-side plugins. In the case of a server-side plugin, the UI would just call some well-known Perses server API to perform those actions and the server plugins could take it from there.

**Reply by Nexucis** [reply]
I think the datasource should be part of the plugin as well. The variable definition can be ( or all the time ) related to the type of the Datasource.

And on Grafana, the plugin is "just" the panel. I assume it's probably because it's super hard or impossible to load at runtime new golang code : https://groups.google.com/g/golang-nuts/c/c269TCEyjWQ

I'm looking at the [CRDs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) because it's an interesting way to add custom object in an API. But in that case there is no interaction with the other object. But it means that we could add more variables definitions dynamically, but the dashboard data definition won't take in consideration these new types.

> Variables: currently in the UI, I'm thinking of this as "how do I get the available values for a variable so a user can make a selection?"

I assume you are talking about different kind of type of variable. But if you are really talking about the actual value of the variables, well it will depend of the kind of the datasource and if it's dynamic ... `GOTO beginning of my comment`.

Otherwise currently you can do like explain in the [doc](https://github.com/perses/perses/blob/master/docs/dashboard.md#how-to-get-the-value-of-the-variables)

Sooooo two solutions from my point of view:

1. if we want to create real plugin with server side, where the plugin is a Golang code like [fluent-bit](https://github.com/fluent/fluent-bit/), we will have to change the language used to implement the backend. Perhaps Rust would be a good candidate (Would like to avoid C, it's a bit too hardcore for me). Quick google search returns interesting result when talking about plugin.
I want to say it's opening some new horizons and we will really have something different than Grafana. Providing a real alternative.

2. Or we do something similar than Grafana:
	* Only Panel can be extendible
	* The backend is providing a sort of proxy that exposed more or less directly the API exposed by the datasource, like `/proxy/datasources/<datasource_kind>/<datasource_name>/<datasource_api like prometheus>`. With that, the logic about how to query the datasource is entirely on the front-end side. (Talking about variables and data for panels). The backend is just modifying the request by injecting any things described in the datasource configuration (like secret, certificat ...etc.)
	* But the dashboard definition will be typeless. I fear we will ended up to have the same kind of dashboard datamodel than Grafana where you don't really get what you are manipulating.
	* Looks like a copy of Grafana ?

**Reply by juliusv** [reply]
Yeah, I neither see us using native Go plugins, nor switching to a different backend language than Go, in my strong opinion. Not only do we all know Go best, but it's also pretty much the standard in the space of people + infra we are working in, and thus we would also be losing a lot of other potential talent and existing library reusability.

So the reality will likely be that only the frontend will be fully pluggable, and while there can be datasource-specific backend code as well, it either has to be statically part of the backend or run as a separate process. So likely a dumb backend datasource proxy makes sense (for people who only want to load new frontend code), at least in addition to a possibly higher abstracted backend proxying method for more well-established and compiled-in datasource plugins in the backend.

**Reply by Nexucis** [reply]
yeah datasource as a plugin could be seen as a webhook like AlertManager. No idea exactly how it could work yet and probably it won't matter so much for the moment.

**Reply by Nexucis** [reply]
mmm if we assert it's impossible or too complicate to do a datasource as a plugin (not even sure it's reallistic to see the datasource through a webhook), probably having the panels plugable doesn't worth it too ?

I mean I get the point of supporting multiple different datasources, but multiple different panels ? The feedback from the SRE is more they wish less panel, because the fancy one are well just fancy but not really necessary. So I'm really wondering where the need of having the panel as a plugin is coming from. Also currently the different panels are typed, so the backend is able to check if the different mandatory fields are properly filled. I think it's important to keep these kind of checks as they will guarantee that your dashboard will be properly displayed.

For sure it can interesting in term of design / architecture to treate the panel as a plugin so it will be easier to add more. But that doesn't necessary mean to open it up for the reason above.

I also get the point on we prefer that the front-end can be able to talk directly to the datasource, For this particular point (which will solve the point 2 and 3), I already have more or less the code to do this kind of thing. I basically just have to transfer it and adapt it a bit. So it should be quick. 

Probably the code managing the build order of the variables can be easily transferable to the frontend as well.

I think at this point of the discussion we can assert that the backend is here to:
* Store the different data like the dashboard, the datsource configuration ...etc.
* A bit later, protect the access to these data using authentication and authorization layer (that will be defined later)
* A proxy to the frontend when it wants to access to the different datasources through `/proxy/datasources/....`. The backend in this case will inject in the request everything required to be able to contact the server (like the certificate / the secret / later finding through the discovery where is the server to contact)

**Comment by LukeTillman** [comment]
One other thing I was thinking (while working on the prototype for plugins in the front-end) was whether or not it's too early to just group front-end plugins into two types like Grafana does (e.g. "Panels" and "DataSource"). For now, what I've done is to keep the concept of a plugin granular. I don't know if that's how it'll end up, but basically how I initially structured things was:

1. **Plugin modules**: a JavaScript bundle that can contain any number of Plugins.
2. **Plugins**: implementations of the actual behavior that's pluggable at a granular level, which are (I think?) going to be made up of React hooks, React components, or both.

You'll see this in my initial PR, but the list of plugin types that you can write a "Plugin" for are:

- `Variables`: how do you load the available options a user can select from for a variable (e.g. run a label names query in Prom, use a static list from config, etc.). This is currently just a hook that you implement to return a list of options (probably from a server), but eventually would include the UI component(s) for editing a variable's options in the Perses UI.
- `ChartQuery`: how do you load data to visualize in a chart (e.g. run an instant or range metrics query against Prom). This is currently just a hook that you implement to return chart data (probably from a server), but eventually would include UI components also for editing the query and its options.
- `Panels`: I want to show something in a Panel on a dashboard. This is currently just a component that you can implement and you get full control over the panel's contents, so it could run a query and graph it, render markup from somewhere (e.g. markdown, etc.) or really whatever you want. Like the others, it would eventually include components for editing the panel's options.

I don't know if thinking of things at this granular of a level is helpful, but I figured it would be good to start granular, and then maybe group them into larger "plugin types" if necessary. While prototyping, I treated a "Data Source" as just a bundle of config that tells me how talk to something (e.g. the base URL for fetching from a Prometheus API). But I'm not sure that's how it should be.

**Reply by LukeTillman** [reply]
Just to elaborate a little bit on why I'm initially thinking this way on two things:

### Keeping plugins fine-grained (at least initially)

One thing that I haven't loved while working in the Grafana plugin API is that it's so coarse-grained (i.e. you either get a Panel or a DataSource), that you end up with an API that you have to implement whether you want to or not. So, you need a "DataSource" that implements a time-series data querying API before you can even think about adding something like a new variable type. Is it possible that a user might want to just implement a new type of variable? Does a variable always need a "Data Source" to fetch data from (e.g. what if the option values are static and just stored in the config)? Are there going to be variables that work in the context of more than one "Data Source" (e.g. does an "Interval" variable make sense in more than just Prometheus)?

Grafana also tries to reuse plugin APIs for multiple things and as a result, the parameters that get passed into them can be pretty gnarly and confusing. For example, the main API for a DataSource plugin (the `.query()` method) is used to run time series data queries for both Dashboard Panels and their Metrics Explorer feature. The options object that gets passed into that contains all sorts of properties to try and cover both use cases, but many times some options don't make sense in one context or the other (e.g. what should the `panelId` be from a Metrics Explorer query or why do you even need it?). To me, it makes sense to try and start fine-grained and use the type system in Typescript to our advantage and to help plugin authors (including us!).

If we ever end up with a similar "Metrics Explorer" type feature, I can imagine that being a separate plugin type (i.e. something like a `MetricsExplorerQuery` plugin vs a `ChartQuery` plugin). There we can define exactly the API that makes sense for that feature and avoid trying to over-generalize and create some perfect abstraction for both (different) use cases. There's nothing stopping plugin authors from reusing core functionality (e.g. how to run a Prom range query, how to do variable templating, etc.) in both plugin implementations. But at least they get an accurate picture (via the type system) of exactly what they need to do to implement a specific Perses feature.

### Front-end plugins are just made up of React Hooks + Components

This one I'm still not so sure on for a number of reasons, but using React primitives as the building blocks of front-end plugins seems like a reasonable starting point. For components, I'm not really sure we have much of an option since Perses is going to be a React application, so the logical/lowest friction way to let plugin authors control how the UI looks is via custom React components. On the hooks side for implementing state/logic, I'm less certain if it's better to let plugins implement a hooks-based API, or just implement plain old functions of some kind. I went with hooks initially because of two main reasons:

1. They give plugin authors a lot of freedom and allow them use of all the React primitives they're used to in their own React applications.
2. We don't have to try and come up with abstractions for a function-based API, particularly around what to pass in as parameters.

I can see number two in particular being an issue. For example, in a `Variable` plugin where you (the plugin author) have to give me a list of options to show, what information does Perses need to pass in as a function parameter to your `getVariableOptions` implementation? Do you need the value of other variables? Do you need the available options of other variables? Do you need the currently selected Dashboard time range? You can imagine scenarios where sometimes you do (e.g. a Prom label names query) and sometimes you don't (e.g. a list of static Intervals to select from). So does that API end up with a bloated list of parameters to cover every possible use case, whether the plugin implementation needs it or not?

I think a hooks-based API solves this problem in a slightly better way in that plugin authors can just use other hooks to get access to the things they need, rather than having to pass "all the things" in via parameters. So if you need dashboard variable values, just `useDashboardVariables()` in your plugin hook and there you have them. If you need the current time range for the dashboard, just `useDashboardTimeRange()` and you're set. In some ways this is kinda getting around the issue by taking advantage of React Context (or Redux, or wherever that state is stored) to get data that's contextual to the hook. The downside to this approach is that I can see it being confusing for a plugin author to know exactly what Perses hooks they can use in a given context.

OK, sorry for the long-winded explanation, just wanted to get some of this out there so you can know where the initial approach came from.
